Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,115 @@
+
+This document exists because Project Saving is a non-trivial challenge to tackle when it comes to ideas with saving and loading interfaces created in both Python and C++ as well as saving and loading of workspaces. This issue should be fairly simple to understand, simply users should be able to save their projects and load them back in, to at least the standard that MantidPlot was able to achieve.
+
+Changes to MantidWorkbench needed
+-----------------------------------------------
+- MainWindow should track all of the open windows (as well as if they are serializable or not) similar to MantidPlot
+- File->Save Project needs to be added
+- File->Load Project needs to be added
+- A GUI for picking where to Save
+
+Options for Implementation of Workspace Saving
+--------------------------------------------------------
+**Python:**
+- Pickling
+	- Pros: Built in library of Python
+	- Cons: Need to add _ _ getstate_ _ and _ _ setstate_ _ functions to every UI class which must return a pickle-able type (dict, list, int, string etc). Then setstate function. Unconfirmed if possible or not.
+- JSON Encoding
+	- Pros: Built in library of Python, "faster" than Pickling: https://www.benfrederickson.com/dont-pickle-your-data/, Potential to be integrated with C++ as it's universal
+	- Cons: Needs methods in every class
+
+**C++**
+- Serialization with boost 
+	- Pros: We already have it, simple, uses an archive type for template serialization, has forward compatibility when classes change.
+	- Cons: We need to give Friend access to each UI class for the boost::serialize::access, not python.
+- Serialization with cereal
+	- Pros: Very easy to implement using a template of Archive type.
+	- Cons: We don't already have the cereal library, plus we'd need to cite it, as per licensing, not python.
+
+**Both C++ and Python**
+- Our own method of serialization utilizing a JSON back-end
+	- Pros: We can pass C++ values to Python based serializer and back to C++, we can determine file format, we can write custom conversions for types (Most Qt types have a QVariant type conversion to JSON anyway), highly flexible
+	- Cons: Requires extra work to implement that a already available library 
+
+
+Implementation of save window, the recommended Method
+-------------------------------------------------------------------------
+Utilizing the serializing method with a JSON back-end, will likely be best as not only will it use a standard file type, it should be easy to change the file type by changing a few functions inside the serializer. Each UI that is serializable should have an X and Y function where X would create a Dictionary with a string as the key and the value set to whatever object you would need to save. Y would use the string key to get back X's value and emplace the original value again, and update the GUI to represent these values.
```

Comment:
The reason I used X and Y is because I had yet to pick a name and wanted some feedback, the issue being that save and load make it sound like it does the actual work of saving and loading and isn't just being called by the save and loading algorithms etc

Reply:
> save and load make it sound like it does the actual work of saving and loading and isn't just being called by the save and loading algorithms

Sounds reasonable, but this would be a good document to suggest a name in. What would you call it instead?

Reply:
Agreed, this is the time to hammer out names etc

Reply:
Perhaps serialize and de-serialize? Although I don't know whether that would be particularly useful as a name because it isn't particularly clear what it will be used for by just glancing at it, but it seems like a more pertinent name considering it must return a dictionary for the documentation to work and isn't really handling the save and/or loading of anything.